### ISSUE TITLE
Other methods of payment

### ISSUE THREAD

**Issue by macjendr** [TrogloGeek/prestashop-tggatos-module#39]
Hi,

great work, it works very well, but I need to ask - is there an easy method to add new methods, i.e. Sprint Secure Solution or 3XWEB from Franfinance? 

Thank you in advance for response :)

**Comment by TrogloGeek** [comment]
Yes, subscribe to SIPS compatible supplementary payment means with your bank and configure accordingly the PAYMENT_MEANS configuration variable from modules back-office

**Comment by macjendr** [comment]
Hello again and thank you for your response, 

Client claims, that Sprint Secure Solution is activated, I have added SOLUTIONSPRINTSECURE,3 payment mean in tggatos configuration and still receiving error 

> Transaction invalide
> 
> Nous regrettons de ne pouvoir donner suite à votre demande, merci de prendre contact directement avec le commercant en indiquant votre problème

Which is different than incorrect payment mean, so it reads sprint secure but something isn't ok. Did I missed something? Shop is already in PRODUCTION mode by the way.

**Comment by TrogloGeek** [comment]
I'll need the documentation provided with this payment mean to further help.
You should ask bank atos support which can tell you what was wrong in your request (just remember to provide them with at least the hour and transaction_id of the request, you can know your transaction ID by activating debug mode in advanced tab)

**Comment by TrogloGeek** [comment]
oh and do not forget this primary rule of web development : NEVER TRUST A CLIENT ;-)

**Comment by macjendr** [comment]
Well, I've received that documentation by email. It's available now here: http://www.pdf-archive.com/2015/10/26/guide-d-implementation-empreinte-cb-uniquement-v0-1/guide-d-implementation-empreinte-cb-uniquement-v0-1.pdf.

I've handled this by adding SOLUTIONSPRINTSECURE_DATA to DATA field in module advanced config with credentials provided by SIPS and it started redirecting correctly, but after that sprintsecure page showing error without any code or clue on google. 

> Le serveur a détecté une erreur
> 
> The server detected an error
> 
> Par sécurité, votre demande a été interrompue. Veuillez éventuellement corriger votre requête et la soumettre à nouveau
> 
> For safety reason, your request was interrupted. Please correct your request if necessary and submit it again

I've wrote to tech support, but no answer for now. Of course I don't want you to waste time on this, unless my mistake is obvious :) It's good to report every step too I think.

**Comment by macjendr** [comment]
> oh and do not forget this primary rule of web development : NEVER TRUST A CLIENT ;-)

Ok, client was mistaken and he have 3XWEB service, so everything works flawlessly!

But one last thing - 3XCBFRANFINANCE_DATA field require full name, address, postal code, city and telephone number of customer. How can I achieve that? For test I have provided dummy info but I can't insert customer variables to DATA field in module config, and I want to avoid module modification if possible.

Thank you in advance.

**Comment by TrogloGeek** [comment]
> > oh and do not forget this primary rule of web development : NEVER TRUST A CLIENT ;-)
> 
> Ok, client was mistaken and he have 3XWEB service, so everything works flawlessly!

Experiences gave me mystical supernatural precognition powers ;-)

> But one last thing - 3XCBFRANFINANCE_DATA field require full name, address, postal code, city and telephone number of customer. How can I achieve that? For test I have provided dummy info but I can't insert customer variables to DATA field in module config, and I want to avoid module modification if possible.

Yes, had the same thought reading the doc you provided. Never worked with this payment mean before nor had a similar case.
It would require one of these paths:
- Make the module aware of this payment mean (splitting PAYMENT_MEANS and searching for its value) and when it is active collect and add required data. And of course make a pull request with your new modules version ;-)
- Order the previous modification from the modules author ;-)
- Call this module from another and use `$mergeParams` parameter of `TggAtos::getPaymentRedirectionForm($amount, Currency $currency, $mode, $mergeParams = array(), $transaction_id = NULL)` to pre-fill the data
- Implement hooks and have another module using hooks to provide additional data

**Comment by macjendr** [comment]
Hello again,

for now I've prepared quick&ugly solution specifically for this case, but I'll be happy to prepare it for pull request and extend this life-saver module with franfinance e-solutions support :)

**Comment by TrogloGeek** [comment]
I'd like to handle this adding a hook system to allow supplementary PrestaShop modules registering themselves against tggatos as sub-modules, providing new payment means and handling special cases.
This would be a neat way to implement this as tggatos would let the sub-module collect the supplementary data and and them to request payload and could as well override/supplement the response decoding system.

This would require several days work and a lot of testing so probably won't happen soon except if many users needs this feature or if one is keen to help developing or financing it.
